src/core/common/uptime.cpp

```diff
@@ -0,0 +1,125 @@
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements mechanism to track device's uptime.
+ */
+
+#include "uptime.hpp"
+
+#if OPENTHREAD_CONFIG_UPTIME_ENABLE
+
+#include "common/code_utils.hpp"
+#include "common/debug.hpp"
+#include "common/instance.hpp"
+#include "common/locator_getters.hpp"
+
+namespace ot {
+
+Uptime::Uptime(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mStartTime(TimerMilli::GetNow())
+    , mOverflowCount(0)
+    , mTimer(aInstance, HandleTimer)
+{
+    mTimer.FireAt(mStartTime + kTimerInterval);
+}
+
+uint64_t Uptime::GetUptime(void) const
+{
+    TimeMilli now           = TimerMilli::GetNow();
+    uint32_t  overflowCount = mOverflowCount;
+
+    // `mTimer` is scheduled with duration `kTimerInterval = (1 << 30)`
+    // so it takes four timer fires for us to go over the entire 32-bit
+    // range and get back to `mStartTime` (at which point we increment
+    // the `mOverflowCount` in `HandleTimer()`).
+    //
+    // Here, we handle the corner case where we may have reached or
+    // passed beyond the `mStartTime` but the `mTimer` is not yet
+    // handled. In this case we increment `overflowCount` (which is
+    // our copy of the current `mOverflowCount` value). We use the
+    // check `(mTimer.GetFireTime() == mStartTime)` to determine if
+    // we are in the last of the four timer fires before overflow.
+
+    if ((mTimer.GetFireTime() == mStartTime) && (now >= mStartTime))
+    {
+        overflowCount++;
+    }
+
+    // The uptime is returned as a `uint64_t`. The `overflowCount`
+    // gives the higher 32 bits, and `(now - mStartTime)` gives the
+    // lower 32 bits (which is always correct even under the corner
+    // case where the `HandleTimer()` is not yet handled).
+
+    return (static_cast<uint64_t>(overflowCount) << 32) + (now - mStartTime);
+}
+
+void Uptime::GetUptime(char *aBuffer, uint16_t aSize) const
+{
+    StringWriter writer(aBuffer, aSize);
+
+    UptimeToString(GetUptime(), writer);
+}
+
+void Uptime::HandleTimer(Timer &aTimer)
+{
+    aTimer.Get<Uptime>().HandleTimer();
+}
+
+void Uptime::HandleTimer(void)
+{
+    if (mTimer.GetFireTime() == mStartTime)
+    {
+        mOverflowCount++;
+    }
+
+    mTimer.FireAt(mTimer.GetFireTime() + kTimerInterval);
+}
+
+void Uptime::UptimeToString(uint64_t aUptime, StringWriter &aWriter)
+{
+    uint64_t days = aUptime / kOneDayInMsec;
+    uint32_t remaining;
+
+    if (days > 0)
+    {
+        aWriter.Append("%lu day%s ", days, (days == 1) ? "" : "s");
```

Comment:
```suggestion
        aWriter.Append("%lud.", days);
```
nit: simpler format can make CLI output parsing easier: `7d.11:22:33:444` v.s. `7 days 11:22:33.444`

Reply:
Good idea. Changed it in new push.